### PR TITLE
util/mpsc: refactor channel

### PR DESCRIPTION
**Pull request by BusyJay**
- remove future support to make performance better
- add half close support
- add several useful APIs

Previous:

```
test channel::bench_channel::bench_crossbeam_channel                  ... bench:          24 ns/iter (+/- 0)
test channel::bench_channel::bench_util_channel                       ... bench:          75 ns/iter (+/- 1)
test channel::bench_channel::bench_util_loose                         ... bench:          90 ns/iter (+/- 5)
```

After:

```
test channel::bench_channel::bench_crossbeam_channel                  ... bench:          27 ns/iter (+/- 0)
test channel::bench_channel::bench_util_channel                       ... bench:          27 ns/iter (+/- 0)
test channel::bench_channel::bench_util_loose                         ... bench:          76 ns/iter (+/- 1)


```

### CODE REVIEW

File: src/util/mpsc.rs [hunk 1]
```diff
@@ -370,42 +418,30 @@ mod tests {
...
+        assert_eq!(tx.len(), 2);
+        assert_eq!(rx.len(), 2);
+        tx.close_sender();
+        assert_eq!(tx.send(3), Err(SendError(3)));
+        assert_eq!(tx.try_send(3), Err(TrySendError::Disconnected(3)));
+        assert!(!tx.is_sender_connected());
+        assert_eq!(rx.try_recv(), Ok(2));
+        assert_eq!(rx.recv(), Ok(3));
+        assert_eq!(rx.try_recv(), Err(TryRecvError::Empty));
+        // assert_eq!(rx.recv(), Err(RecvError));
```
Comment (overvenus) [inline src/util/mpsc.rs]: Uncomment?

File: src/util/mpsc.rs [hunk 2]
```diff
@@ -77,97 +76,115 @@ impl<T> Clone for Sender<T> {
...
+    }
+
+    /// Blocks the current thread until a message is sent or the channel is disconnected.
     #[inline]
     pub fn send(&self, t: T) -> Result<(), SendError<T>> {
-        if !self.state.is_receiver_closed() {
-            self.sender.send(t);
-            self.notify();
-            return Ok(());
+        if self.state.is_sender_connected() {
```
Comment (hicqu) [inline src/util/mpsc.rs]: The latest `crossbeam::channel::Sender::send` returns `Result` now, so how about remove the `state` at all?

File: src/util/mpsc.rs (hunk 1 above)
Comment (BusyJay) [reply src/util/mpsc.rs]: It will be uncommented when full close is implemented.

File: src/util/mpsc.rs (hunk 2 above)
Comment (BusyJay) [reply src/util/mpsc.rs]: `State` is still useful for close sender.

File: src/util/mpsc.rs [hunk 3]
```diff
@@ -190,76 +208,87 @@ pub fn bounded<T>(cap: usize) -> (Sender<T>, Receiver<T>) {
...
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.sender.is_empty()
+    }
+
     /// Send a message regardless its capacity limit.
     #[inline]
-    pub fn force_send(&mut self, t: T) -> Result<(), SendError<T>> {
-        self.tried_cnt += 1;
+    pub fn force_send(&self, t: T) -> Result<(), SendError<T>> {
```
Comment (hicqu) [inline src/util/mpsc.rs]: Why not `&mut self`? I think it's the common style. And `Cell` make the `Sender` not `Sync` any more.

Comment (BusyJay) [reply src/util/mpsc.rs]: It's intended. Actually, `&self` is very common for sending a message. And not using a mutable reference is useful for integrating it into existing transport system.

File: src/util/mpsc.rs (hunk 2 above)
Comment (hicqu) [reply src/util/mpsc.rs]: Got it.

File: src/util/mpsc.rs (hunk 3 above)
Comment (hicqu) [reply src/util/mpsc.rs]: Ok. So do you think `impl !Sync for LooseBoundedSender` explicitly is a good idea? Just make the doc more friendly.

Comment (BusyJay) [reply src/util/mpsc.rs]: That would be unnecessary. `!Sync` is not intended. Internal mutability is intended. It's a quite common in rust world that sender is not Sync as sender in std library is not Sync.

**Pull request merged**